### PR TITLE
Fix bug with managing redirection and keepAuthorization in Web cmdlets

### PR DESCRIPTION
**Pull request by CarloToso**
# PR Summary

The current behaviour of `if (keepAuthorization && IsRedirectCode(response.StatusCode) && response.Headers.Location is not null)` is wrong, I tested it by setting `keepAuthorization = true` --> this sets `handleRetirect = true` --> and then sets `handler.AllowAutoRedirect = false`. Without `sessionRedirect` (we can choose a better name) it ignores the `MaximumRedirection` parameter.

Test: maunally set `keepAuthorization = true` before the `if`, then
`Invoke-WebRequest "http://mockbin.org/redirect/308?to=https://mockbin.org/redirect/302?to=https://mockbin.org/status/200" -MaximumRedirection 0 -SkipHttpErrorCheck` --> expected 308, result 200

Proposed fix --> 
```cs
bool sessionRedirect = WebSession.MaximumRedirection > 0 || WebSession.MaximumRedirection == -1;
if (keepAuthorization  && sessionRedirect && IsRedirectCode(response.StatusCode) && response.Headers.Location is not null)
```

## PR Context

Fix Bug, merge before #18546 and #18894

### CODE REVIEW

**Comment by iSazonov** [comment]
@CarloToso Can you enhance our helper test tool - WebListener - to cover the scenario? Currently it supports redirection but not take into account authorization header (I guess we could use `-Authentication Bear`?). WebListener could return the auth token as-is and we could add tests to check that the header is removed by default and isn't removed with `PreserveAuthorizationOnRedirect`.

I see only reason we do custom redirection processing is that we want to keep the auth header. So I suggest rename `keepAuthorization` and `handleRedirect` parameters to `keepAuthorizationOnRedirect`. This simplify understanding the code.

**Comment by CarloToso** [comment]
@iSazonov we also need to handle redirection for #18546 and #18894, I can change `keepAuthorization` to `keepAuthorizationOnRedirect` but i think `handleRedirect` should stay as it is

**Comment by iSazonov** [comment]
> but i think `handleRedirect` should stay as it is.

It is the same as `keepAuthorization` - we assign keepAuthorization to handleRedirect - this mislead.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs [hunk 1]
```diff
@@ -1313,7 +1313,9 @@ internal virtual HttpResponseMessage GetResponse(HttpClient client, HttpRequestM
                 _cancelToken = new CancellationTokenSource();
                 response = client.SendAsync(req, HttpCompletionOption.ResponseHeadersRead, _cancelToken.Token).GetAwaiter().GetResult();
 
-                if (keepAuthorization && IsRedirectCode(response.StatusCode) && response.Headers.Location is not null)
+                bool sessionRedirect = WebSession.MaximumRedirection > 0 || WebSession.MaximumRedirection == -1;
+
+                if (keepAuthorization && sessionRedirect && IsRedirectCode(response.StatusCode) && response.Headers.Location is not null)
```
Comment (xtqqczze) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs:1318]:
```suggestion
                if (keepAuthorization
                    && WebSession.MaximumRedirection is not 0
                    && IsRedirectCode(response.StatusCode)
                    && response.Headers.Location is not null)
```

**Comment by CarloToso** [comment]
If we change it handleRedirect to keepAuthorization, it will be misleading after we add -PreserveHTTPMethodOnRedirect or -AllowInsicureRedirect. We could add something like this `bool handleRedirect = keepAuthorizationOnRedirect || AllowInsicureRedirect || PreserveHTTPMethodOnRedirect`

**Comment by iSazonov** [comment]
> If we change it handleRedirect to keepAuthorization, it will be misleading after we add -PreserveHTTPMethodOnRedirect or -AllowInsicureRedirect. We could add something like this `bool handleRedirect = keepAuthorizationOnRedirect || AllowInsicureRedirect || PreserveHTTPMethodOnRedirect`

I don't see a problem here. As I said only reason we do manual redirect processing is that we want to keep auth header. So variable names should reflect to our intention. And comments could be updated too.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs [hunk 2]
```diff
@@ -533,7 +533,7 @@ internal virtual void ValidateParameters()
             }
 
             // Resume requires OutFile.
-            if (Resume.IsPresent && OutFile is null)
+            if (Resume && OutFile is null)
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs:536]: This reduce readability. 😕

Comment (CarloToso) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs]: reverted

**Comment by CarloToso** [comment]
@iSazonov It looks like there already are some tests that cover -PreserveAuthorizationOnRedirect

**Comment by iSazonov** [comment]
> @iSazonov It looks like there already are some tests that cover -PreserveAuthorizationOnRedirect

Can you point these tests? I see there is Authorization controller but it doesn't cover redirection scenario.

**Comment by iSazonov** [comment]
If the test is passed before the PR it is a broken test.

**Comment by CarloToso** [comment]
I think the tests are not broken, they just never checked for `-MaximumRedirection` + `-PreserveAuthorizationOnRedirect`

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs [hunk 3]
```diff
@@ -1451,12 +1447,14 @@ protected override void ProcessRecord()
 
                 // if the request contains an authorization header and PreserveAuthorizationOnRedirect is not set,
                 // it needs to be stripped on the first redirect.
-                bool keepAuthorization = WebSession is not null &&
-                                         WebSession.Headers is not null &&
-                                         PreserveAuthorizationOnRedirect.IsPresent &&
-                                         WebSession.Headers.ContainsKey(HttpKnownHeaderNames.Authorization);
+                bool keepAuthorizationOnRedirect = WebSession is not null &&
+                                                   WebSession.Headers is not null &&
+                                                   PreserveAuthorizationOnRedirect &&
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs:1452]:
```suggestion
                                                   PreserveAuthorizationOnRedirect.IsPresent &&
```

File: test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1 [hunk 4]
```diff
@@ -133,14 +133,22 @@ function ExecuteRedirectRequest {
...
+        [int]
+        $MaximumRedirection
     )
     $result = [PSObject]@{Output = $null; Error = $null; Content = $null}
 
     try {
         $headers = @{"Authorization" = "test"}
         if ($Cmdlet -eq 'Invoke-WebRequest') {
-            $result.Output = Invoke-WebRequest -Uri $uri -Headers $headers -PreserveAuthorizationOnRedirect:$PreserveAuthorizationOnRedirect.IsPresent -Method $Method
+            if ($MaximumRedirection) {
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1:147]:
```suggestion
            if ($MaximumRedirection -gt 0) {
```

**Review by iSazonov — COMMENTED** [review]
Please confirm that the new tests fail before the PR.

**Comment by CarloToso** [comment]
The new test fails before this PR:
Immagine 2023-01-12 100146
And passes after:
Immagine 2023-01-12 100945

**Review by iSazonov — APPROVED** [review]

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs [hunk 5]
```diff
@@ -1451,12 +1447,14 @@ protected override void ProcessRecord()
...
-                bool keepAuthorization = WebSession is not null &&
-                                         WebSession.Headers is not null &&
-                                         PreserveAuthorizationOnRedirect.IsPresent &&
-                                         WebSession.Headers.ContainsKey(HttpKnownHeaderNames.Authorization);
+                bool keepAuthorizationOnRedirect = WebSession is not null &&
+                                                   WebSession.Headers is not null &&
+                                                   PreserveAuthorizationOnRedirect.IsPresent &&
+                                                   WebSession.Headers.ContainsKey(HttpKnownHeaderNames.Authorization);
+
+                bool handleRedirect = keepAuthorizationOnRedirect; // || AllowInsicureRedirect || PreserveHTTPMethodOnRedirect
```
Comment (xtqqczze) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs:1455]: Remove or refactor the comment.

**Review by xtqqczze — CHANGES_REQUESTED** [review]

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs (hunk 5 above)
Comment (CarloToso) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs]: Removed comment

**Comment by iSazonov** [comment]
@CarloToso Please resolve merge conflicts.

**Comment by iSazonov** [comment]
@CarloToso Please resolve merge conflict again.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs [hunk 6]
```diff
@@ -1418,14 +1414,16 @@ protected override void ProcessRecord()
...
                 // it needs to be stripped on the first redirect.
-                bool keepAuthorization = WebSession is not null
-                                         && WebSession.Headers is not null
-                                         && PreserveAuthorizationOnRedirect.IsPresent
-                                         && WebSession.Headers.ContainsKey(HttpKnownHeaderNames.Authorization);
+                bool keepAuthorizationOnRedirect = WebSession is not null
+                                                   && WebSession.Headers is not null
+                                                   && PreserveAuthorizationOnRedirect.IsPresent
+                                                   && WebSession.Headers.ContainsKey(HttpKnownHeaderNames.Authorization);
+
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs:1423]: It seems `PreserveAuthorizationOnRedirect.IsPresent` should be on first place.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs [hunk 7]
```diff
@@ -1282,7 +1275,10 @@ internal virtual HttpResponseMessage GetResponse(HttpClient client, HttpRequestM
                 _cancelToken = new CancellationTokenSource();
                 response = client.SendAsync(req, HttpCompletionOption.ResponseHeadersRead, _cancelToken.Token).GetAwaiter().GetResult();
 
-                if ((keepAuthorization || (AllowInsecureRedirect && (WebSession.MaximumRedirection > 0 || WebSession.MaximumRedirection == -1))) && IsRedirectCode(response.StatusCode) && response.Headers.Location != null)
+                if (handleRedirect
+                    && WebSession.MaximumRedirection is not 0
+                    && IsRedirectCode(response.StatusCode)
+                    && response.Headers.Location is not null)
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs:1281]: Why is AllowInsecureRedirect removed?

We merged #18546 without tests. Time to add tests for AllowInsecureRedirect

**Review by iSazonov — CHANGES_REQUESTED** [review]

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs (hunk 7 above)
Comment (CarloToso) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs:1281]: Added some tests #18939

**Comment by iSazonov** [comment]
@CarloToso Please rebase to get tests.

**Review by iSazonov — APPROVED** [review]

**Pull request merged**